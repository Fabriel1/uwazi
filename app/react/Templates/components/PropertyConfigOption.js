import { Field } from 'react-redux-form';
import PropTypes from 'prop-types';
import React from 'react';

const PropertyConfigOption = ({ children, model, label }) => (
  <Field model={model}>
    <label className="property-label" htmlFor={`test${model}`}>
      <input id={`test${model}`} type="checkbox"/> {label}
    </label>
    {children}
  </Field>
);

PropertyConfigOption.defaultProps = {
  children: ''
};

PropertyConfigOption.propTypes = {
  children: PropTypes.node,
  model: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired
};

export default PropertyConfigOption;
