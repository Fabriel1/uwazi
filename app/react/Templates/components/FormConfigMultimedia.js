import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Field } from 'react-redux-form';
import { connect } from 'react-redux';

import { t } from 'app/I18N';
import { Select } from 'app/ReactReduxForms';
import Tip from 'app/Layout/Tip';

import PropertyConfigOption from './PropertyConfigOption';

const style = index => (
  <div>
    <div className="form-group">
      <label>
        {t('System', 'Style')}
      </label>
      <Select
        model={`template.data.properties[${index}].style`}
        options={[{ _id: 'contain', name: 'Contain' }, { _id: 'cover', name: 'Cover' }]}
        optionsLabel="name"
        optionsValue="_id"
      />
    </div>
    <div className="protip">
      <p>
        <b>Contain</b> will show the entire media inside the container.  Grey bars may appear to the sides of the image.
        <br /><b>Cover</b> will attempt to fill the container, using it&#39;s entire width.  Cropping is likely to occur.
      </p>
    </div>
  </div>
);

class FormConfigMultimedia extends Component {
  render() {
    const { property, index, formState, canShowInCard, helpText } = this.props;

    let labelClass = 'form-group';
    const labelKey = `properties.${index}.label`;
    const requiredLabel = formState.$form.errors[`${labelKey}.required`];
    const duplicatedLabel = formState.$form.errors[`${labelKey}.duplicated`];
    if (requiredLabel || duplicatedLabel) {
      labelClass += ' has-error';
    }

    return (
      <div>
        <div className={labelClass}>
          <label>Name</label>
          <Field model={`template.data.properties[${index}].label`}>
            <input className="form-control"/>
          </Field>
        </div>

        {helpText &&
          <div className="protip">
            <i className="fa fa-lightbulb-o"/>
            <span>{helpText}</span>
          </div>
        }

        <PropertyConfigOption label="Show label" model={`template.data.properties[${index}].showLabel`}>
          <Tip>If selected, the property will be rendered with the label. Otherwise, it will just show the multimedia content.</Tip>
        </PropertyConfigOption>

        {canShowInCard &&
          <div>
            <PropertyConfigOption label="Show in cards" model={`template.data.properties[${index}].showInCard`}>
              <Tip>This property will appear in the library cards as part of the basic info.</Tip>
            </PropertyConfigOption>
            {property.showInCard && style(index)}
          </div>
        }
      </div>
    );
  }
}

FormConfigMultimedia.defaultProps = {
  canShowInCard: true,
  helpText: ''
};

FormConfigMultimedia.propTypes = {
  property: PropTypes.instanceOf(Object).isRequired,
  canShowInCard: PropTypes.bool,
  helpText: PropTypes.string,
  index: PropTypes.number.isRequired,
  formState: PropTypes.instanceOf(Object).isRequired
};

export function mapStateToProps({ template }, ownProps) {
  return {
    property: template.data.properties[ownProps.index],
    formState: template.formState
  };
}

export default connect(mapStateToProps)(FormConfigMultimedia);
