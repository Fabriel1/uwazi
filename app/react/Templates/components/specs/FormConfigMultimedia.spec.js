import React from 'react';
import { shallow } from 'enzyme';
import { Field } from 'react-redux-form';

import FormConfigMultimedia, { mapStateToProps } from '../FormConfigMultimedia';
import PropertyConfigOption from '../PropertyConfigOption';

describe('FormConfigMultimedia', () => {
  let component;
  let state;
  let props;

  beforeEach(() => {
    props = { index: 0 };
    state = {
      template: {
        data: { properties: [{ label: '' }] },
        formState: {
          'properties.0.label': { valid: true, dirty: false, errors: {} },
          $form: {
            errors: {
              'properties.0.label.required': false,
              'properties.0.label.duplicated': false
            }
          }
        }
      }
    };
  });

  const render = () => {
    const mappedProps = Object.assign({}, props, mapStateToProps(state, props));
    component = shallow(<FormConfigMultimedia.WrappedComponent {...mappedProps}/>);
  };

  const expectErrorLengthToBe = (length) => {
    render();
    expect(component.find('.has-error').length).toBe(length);
  };

  it('should hold show label, show in card and other congifuration options by default', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  it('should add option to select card display type', () => {
    state.template.data.properties[0].showInCard = true;
    render();
    expect(component).toMatchSnapshot();
  });

  it('should allow setting a help text', () => {
    render();
    expect(component.find('.protip').length).toBe(0);
    props.helpText = 'Some help text';
    render();
    expect(component).toMatchSnapshot();
  });

  it('should allow excluding "show in card"', () => {
    props.canShowInCard = false;
    render();
    const formFields = component.find(Field);
    expect(formFields.getElements().length).toBe(1);
  });

  describe('validation', () => {
    it('should render the label without errors', () => {
      expectErrorLengthToBe(0);
    });
  });

  describe('when the field is invalid and dirty or the form is submited', () => {
    it('should render the label with errors', () => {
      state.template.formState.$form.errors['properties.0.label.required'] = true;
      state.template.formState['properties.0.label'].dirty = true;
      expectErrorLengthToBe(1);
    });

    it('should render the label with errors', () => {
      state.template.formState.$form.errors['properties.0.label.required'] = true;
      state.template.formState.submitFailed = true;
      expectErrorLengthToBe(1);
    });
  });
});
