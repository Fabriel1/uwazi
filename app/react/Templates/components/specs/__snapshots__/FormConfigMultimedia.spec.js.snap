// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormConfigMultimedia should allow setting a help text 1`] = `
<div>
  <div
    className="form-group"
  >
    <label>
      Name
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
      />
    </Modeled(Connect(Field))>
  </div>
  <div
    className="protip"
  >
    <i
      className="fa fa-lightbulb-o"
    />
    <span>
      Some help text
    </span>
  </div>
  <PropertyConfigOption
    label="Show label"
    model="template.data.properties[0].showLabel"
  >
    <Tip>
      If selected, the property will be rendered with the label. Otherwise, it will just show the multimedia content.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[0].showInCard"
  >
    <Tip>
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div>
    <div
      className="form-group"
    >
      <label>
        Style
      </label>
      <Select
        model="template.data.properties[0].style"
        options={
          Array [
            Object {
              "_id": "contain",
              "name": "Contain",
            },
            Object {
              "_id": "cover",
              "name": "Cover",
            },
          ]
        }
        optionsLabel="name"
        optionsValue="_id"
      />
    </div>
    <div
      className="protip"
    >
      <p>
        Style mostly affects cards, but has minor effects in side panels and entity views.
      </p>
      <p>
        <b>
          Contain
        </b>
         will show the entire media inside the container.  Grey bars may appear to the sides of the image.
        <br />
        <b>
          Cover
        </b>
         will attempt to fill the container, using it's entire width.  Cropping is likely to occur.
      </p>
    </div>
  </div>
</div>
`;

exports[`FormConfigMultimedia should hold show label, show in card and card display type options by default 1`] = `
<div>
  <div
    className="form-group"
  >
    <label>
      Name
    </label>
    <Modeled(Connect(Field))
      model="template.data.properties[0].label"
    >
      <input
        className="form-control"
      />
    </Modeled(Connect(Field))>
  </div>
  <PropertyConfigOption
    label="Show label"
    model="template.data.properties[0].showLabel"
  >
    <Tip>
      If selected, the property will be rendered with the label. Otherwise, it will just show the multimedia content.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[0].showInCard"
  >
    <Tip>
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div>
    <div
      className="form-group"
    >
      <label>
        Style
      </label>
      <Select
        model="template.data.properties[0].style"
        options={
          Array [
            Object {
              "_id": "contain",
              "name": "Contain",
            },
            Object {
              "_id": "cover",
              "name": "Cover",
            },
          ]
        }
        optionsLabel="name"
        optionsValue="_id"
      />
    </div>
    <div
      className="protip"
    >
      <p>
        Style mostly affects cards, but has minor effects in side panels and entity views.
      </p>
      <p>
        <b>
          Contain
        </b>
         will show the entire media inside the container.  Grey bars may appear to the sides of the image.
        <br />
        <b>
          Cover
        </b>
         will attempt to fill the container, using it's entire width.  Cropping is likely to occur.
      </p>
    </div>
  </div>
</div>
`;
